File: common/src/main/java/cc/turtl/chiselmon/util/ParseUtils.java
package cc.turtl.chiselmon.util;

import java.text.Normalizer;
import java.util.Locale;
import java.util.regex.Pattern;

public class ParseUtils {
    private static final Pattern DIACRITICS = Pattern.compile("\\p{M}");
    private static final Pattern NON_ALPHANUMERIC = Pattern.compile("[^a-z0-9]");

    public static String normalizeSpeciesName(String input) {
        if (input == null || input.isBlank()) {
            return "";
        }

        // Normalize (Decompose characters like 'é' to 'e' + accent)
        String normalized = Normalizer.normalize(input, Normalizer.Form.NFD);

        // Remove accents
        String stripped = DIACRITICS.matcher(normalized).replaceAll("");

        // Lowercase and strip everything except a-z and 0-9
        return NON_ALPHANUMERIC.matcher(stripped.toLowerCase(Locale.ROOT)).replaceAll("");
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/util/ParseUtils.kt
package cc.turtl.chiselmon.core.util

import java.text.Normalizer

private val DIACRITICS = Regex("\\p{M}")
private val NON_ALPHANUMERIC = Regex("[^a-z0-9]")

fun normalizeSpeciesName(input: String?): String {
    if (input.isNullOrBlank()) return ""

    // Decompose characters like 'é' into base letter + combining accent, then strip accents
    val stripped = DIACRITICS.replace(Normalizer.normalize(input, Normalizer.Form.NFD), "")

    return NON_ALPHANUMERIC.replace(stripped.lowercase(), "")
}
